projects/storefrontlib/src/cms-structure/services/feature-modules.service.ts

```diff
@@ -0,0 +1,254 @@
+import {
+  Compiler,
+  Injectable,
+  InjectFlags,
+  Injector,
+  NgModuleFactory,
+  NgModuleRef,
+  OnDestroy,
+} from '@angular/core';
+import {
+  CMSComponentConfig,
+  CmsComponentMapping,
+  CmsConfig,
+  ConfigChunk,
+  ConfigInitializerService,
+  configurationFactory,
+  DefaultConfigChunk,
+  FeatureModuleConfig,
+} from '@spartacus/core';
+import {
+  combineLatest,
+  defer,
+  forkJoin,
+  from,
+  Observable,
+  of,
+  queueScheduler,
+} from 'rxjs';
+import { map, observeOn, pluck, share, switchMap, tap } from 'rxjs/operators';
+
+interface FeatureInstance extends FeatureModuleConfig {
+  moduleRef?: NgModuleRef<any>;
+  depsModules?: any[];
+  injectors?: Injector[];
+  componentsMappings?: CMSComponentConfig;
+}
+
+/**
+ * Service responsible for resolving feature domains
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class FeatureModulesService implements OnDestroy {
+  // maps componentType to feature
+  private componentFeatureMap: Map<string, string> = new Map();
+
+  private featureModulesConfig?: {
+    [featureName: string]: FeatureInstance;
+  };
+
+  private featureResolvers: Map<
+    string,
+    Observable<FeatureInstance>
+  > = new Map();
+
+  private features: Map<string, FeatureInstance> = new Map();
+
+  private dependencyModules = new Map<any, NgModuleRef<any>>();
```

Comment:
What if some feature module is a dependency for a different feature module? Will it be resolved twice?

Reply:
Yes, but it should not be a case. 
Another example: we can have two features, that are using the same Anguar module: they will be resolved (and instantiated) twice.
Dependency modules will only be instantiated once.